crates/items/src/item_stack.rs

```diff
@@ -0,0 +1,262 @@
+        self.item == other.item && self.has_same_damage(other)
+    }
+
+    /// Returns the item type for this `ItemStack`.
+    pub fn item(&self) -> Item {
+        self.item
+    }
+
+    /// Returns the number of items in this `ItemStack`.
+    pub fn count(&self) -> u32 {
+        self.count
+    }
+
+    /// Adds more items to this `ItemStack`. Returns the new count.
+    pub fn add(&mut self, count: u32) -> Result<u32, ItemStackError> {
+        self.set_count(self.count + count)
+    }
+
+    /// Adds more items to this `ItemStack`. Does not check if the
+    /// addition will make the count to be greater than the
+    /// stack size. Does not check count overflows. Returns the new count.
+    pub fn unchecked_add(&mut self, count: u32) -> u32 {
+        self.count += count;
+        self.count
+    }
+
+    /// Removes some items from this `ItemStack`.
+    pub fn remove(&mut self, count: u32) -> Result<u32, ItemStackError> {
+        if self.count < count {
+            return Err(ItemStackError::NotEnoughAmount);
+        }
+        self.count -= count;
+        Ok(self.count)
+    }
+
+    /// Sets the item type for this `ItemStack`. Returns the new
+    /// item type or fails if the current item count exceeds the
+    /// new item type stack size.
+    pub fn set_item(&mut self, item: Item) -> Result<Item, ItemStackError> {
+        if self.count > item.stack_size() {
+            return Err(ItemStackError::ExceedsStackSize);
+        }
+        self.item = item;
+        Ok(self.item)
+    }
+
+    /// Sets the item type for this `ItemStack`. Does not check if
+    /// the new item type stack size will be lower than the current
+    /// item count. Returns the new item type.
+    pub fn unchecked_set_item(&mut self, item: Item) -> Item {
+        self.item = item;
+        self.item
+    }
+
+    /// Sets the count for this `ItemStack`. Returns the updated
+    /// count or fails if the new count would exceed the stack
+    /// size for that item type.
+    pub fn set_count(&mut self, count: u32) -> Result<u32, ItemStackError> {
+        if count > self.item.stack_size() {
+            return Err(ItemStackError::ExceedsStackSize);
+        } else if count > i32::MAX as u32 {
+            return Err(ItemStackError::ClientOverflow);
+        }
+        self.count = count;
+        Ok(self.count)
+    }
+
+    /// Sets the count for this `ItemStack`. It will not check if
+    /// the desired count exceeds the current item type stack size.
+    ///  Does not check count overflows. Returns the updated count.
+    pub fn unchecked_set_count(&mut self, count: u32) -> u32 {
+        self.count = count;
+        self.count
+    }
+
+    /// Splits this `ItemStack` in half, returning the
+    /// removed half. If the amount is odd, `self`
+    /// will be left with the least items. Returns the taken
+    /// half.
+    pub fn take_half(&mut self) -> ItemStack {
+        self.take((self.count as f64 / 2 as f64).ceil() as u32)
```

Comment:
I'm always suspicious of using floating-point operations for integer arithmetic, but I might just be pedantic. This can be implemented as `((self.count + 1) / 2)` instead.

Reply:
Well, `(self.count + 1) / 2` is more concise and efficient. Thanks 😄